Patch: Uncertainties

evandromr opened the pull request:
This PR add uncertainties atributes to Lightcurve, Powerspectrum and AveragedPowerspectrum. As discussed in Issue #23 

For the light curve it is assumed that the data follows a Poisson distribution (which is the case for most Astrophysical cases), therefore the standard deviation is the square root of the number of counts.
For count rate the square root is divided by the time interval in which the rate is averaged. 
I believe this is correct in general.

For a single powerspectrum the uncertainties is equal to the power itself. For an averaged (and/or binned) powerspectrum the uncertainties is equal to the power divided by the square root of the number of averaged segments (and/or frequency bins).

I believe this is generally correct for a Leahy normalized powerspectrum, but probably not for an RMS normalized one or at least I'm not sure. This should be checked before merged. Any thoughts on this?

I added small descriptions to the doc strings and one comment in the body of the code.
Feel free to propose any changes.

I prefer the term "uncertainty" but I called the variables counts_err, ps_err, and so on because I believe most people are used to call it "error" and its easier to associate to error bars when plotting.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/stingraysoftware/stingray/24)
<!-- Reviewable:end -->

dhuppenkothen commented:
I think your definition of the uncertainties should be correct even in rms normalization.

I haven't forgotten this, just writing proposals this week. I'll get back to review this ASAP after Friday.

evandromr commented:
I incorporated in this PR a fix for a bug in the `rebin` method.

The atribute `m` in the output of the `rebin` method takes into account the atribute `self.m` of the input spectra. This is important when re-bining averaged powerspectra since the number of realizations for each power should be `m = number_of_powers_in_a_bin * number_of_spectra_averaged`.

This seems to agree with [van der Klis chapter](http://link.springer.com/chapter/10.1007%2F978-94-009-2273-0_3) in *"Fourier Techniques in X-Ray Timing"*

dhuppenkothen commented:
Okay, this looks good.

However, before merging, I'm wondering whether we really want to force Lightcurve to assume Poisson data. This seems unnecessarily restrictive to anyone who might want to put in a Lightcurve that's, say, optical data. 

My suggestion would be to make a subclass PoissonLightcurve whose __init__ calls the Lightcurve __init__ (without the errors on the counts), and then defines the errors on counts and count rate there. 

One could then imagine that we might add a GaussianLightcurve subclass in the future that might have its own keyword argument for the standard deviation of the data? 

Does this sound like it would make sense or is it unnecessarily complicated?

evandromr commented:
You make a good point. However I think that creating 2 different classes will make it too complicated for a clueless user.

I believe that in most cases where you wouldn't use Poisson statistics, you have the uncertainties along with the countrate data. And besides that, thinking about it now, the user should have the freedom to input his/her own array of uncertainties. Even for optical lightcurves, if the light curve is created from the "toa" of photons the statistic is Poisson but approaches a Gaussian distribution. If we don't have events I don't see how we could calculate non-poissonic standard error for already binned data.

My solution would be adding an optional argument to Lightcurve ("`err=None`") that, if given,  is used as the `counts_err` (or `countrate_err`, depending on the value of `input_counts`), and if not it is assumed Poisson statistic. 

This would require modification to the `rebin` method, to also rebin the uncertainties instead of assuming Poisson from the rebinned counts array.

evandromr commented:
To not mess with the original proposal of the PR, I created another branch `patch_uncertainties2` with this new proposal. You can check if it makes sense. If so, I'll merge the new branch in this PR

dhuppenkothen commented:
Yes, I like that much better. I had to think about it for a while, and I think you're right. Poisson statistics is not a bad default assumption if there are no errors, and if it's not Poisson, the user should supply those errors. Sometimes I try to be more general than I have to. :)

Feel free to merge this branch into your pull request, and I'll have a look at the entire thing.

dhuppenkothen commented:
PS: Can you also make a "Contributers" section under "Authors" in the README and add yourself?

PS2: We were wondering whether you'd like to contribute regularly and want write access to the repo? Submission of new code is still via pull request, but this would also give you the chance to review the code changes we're working on and comment on/merge those. We'd love to have you on board. :)

evandromr commented:
Hi, I updated the implementation to take the keyword `err` in `Lightcurve` and to use an `yerr` variable in the utils.rebin_data() function.

I also add myself as "Contributer" in the README, as you suggested. And yes, I would like to contribute more :)
I believe you are also joining the Workshop in Leiden next week (I saw your name in the list of participants), we can talk more about it there.

dhuppenkothen commented:
Looks good! Last nitpick from me: Could you either put a `TODO` in the relevant methods that this code requires tests, or write a few tests to show that the code does the right thing? 

I'm happy to merge and write a few tests myself soon, but I'd like a reminder somewhere that it needs to be done.

evandromr commented:
Yes, I forgot to mention the lack of tests. I'll do it as soon as I got enough the time.

evandromr commented:
Added tests for the uncertainties in light curves and powerspectra, and updated the fork to latest state of branch `master`

Sorry for the messy-ish commit history. I should have kept some of the aesthetic changes out of the PR :no_mouth:

dhuppenkothen commented on stingray/lightcurve.py:
```diff
@@ -70,9 +81,25 @@ def __init__(self, time, counts, input_counts=True):
             self.countrate = self.counts/self.dt
+            if err is not None:
+                assert np.all(np.isfinite(err)), "There are inf or NaN " \
+                                                 "values in your err array!"
```
I think you should be able to pull that assertion out of the two if statements to save  you the duplication.

evandromr replied:
I tried, but the tests failed and `pytest` didn't caught `TypeError` as it should.
I will play with it a little more to see if I find the bug.

dhuppenkothen commented on stingray/powerspectrum.py:
```diff
@@ -272,9 +279,11 @@ def rebin(self, df, method="mean"):
+                                                        self.ps[1:],
+                                                        self.ps_err[1:], df,
+                                                        method=method)
 
```
Have you pulled the latest version of the code? I think with the new definition of the Fourier frequencies, the arguments should just be `self.freq`, `self.ps` and `self.ps_err`. Leaving out the first element of the array was only useful when the zeroth frequency was still part of the array.

evandromr replied:
Yes, it was pulled from the latest commit to `master`. Is it wrong upstream then?

evandromr replied:
I'm now confused on what is supposed to be right behavior. I had to change the test function `rebin_data`to pass the tests with the changes tha you suggested.

Can you please check the latest build?

dhuppenkothen commented:
Yes, that's wrong upstream. It might mean when you change it some of the tests fail. We'll see.

matteobachetti commented:
@evandromr please check the conflicts :)

evandromr commented on stingray/tests/test_powerspectrum.py:
```diff
@@ -184,7 +181,7 @@ def rebin_several(self, df):
         ps = Powerspectrum(lc=self.lc, norm="Leahy")
         bin_ps = ps.rebin(df)
-        assert np.isclose(bin_ps.freq[0], bin_ps.df, atol=1e-4, rtol=1e-4)
+        assert np.isclose(bin_ps.freq[1], bin_ps.df, atol=1e-4, rtol=1e-4)
```
I changed the rebin method on powerspectrum as @dhuppenkothen suggested but the tests failed then. I changed the assertion test in the function `tests.test_powerspectrum.rebin_several()'

Can someone else, please, check what is the expected behavior with the new frequency array?

evandromr commented:
Just updating the thread: This PR depends on #95 (with discussion on #82 ).

Also the commit history is quite dirty by now. Should we close this one and I open a new PR when/if #95 is accepted?

nithinsingh61 commented on stingray/lightcurve.py:
```diff
@@ -56,30 +68,36 @@ def __init__(self, time, counts, input_counts=True):
+            assert np.all(np.isfinite(err)), "There are inf or NaN " \
+                                             "values in your err array!"
+        else:
+            err = np.sqrt(np.asarray(counts))
```
I think we should add another layer of refinement when the counts per bin are very small (counts  < ~20-30)
`error = 1 + sqrt(N + 0.75)`  (the Gehrels approximation [link](http://articles.adsabs.harvard.edu/cgi-bin/nph-iarticle_query?1986ApJ...303..336G&amp;data_type=PDF_HIGH&amp;whole_paper=YES&amp;type=PRINTER&amp;filetype=.pdf), [link2](http://stsdas.stsci.edu/cgi-bin/gethelp.cgi?explain_errors) )

dhuppenkothen commented:
ping @evandromr: I don't know what the status of this PR is. Did we ever implement uncertainties elsewhere, in which case this can be closed? Or does this still need implementation?

evandromr commented:
@dhuppenkothen As far as I know neither Lightcurve nor Crosspectrum has uncertainties implemented yet. I stopped the development of this PR when I noticed that the binning of Lightcurves was not as I expected.

I opened Issue #82 to discuss the `rebin()` method but nobody has commented on it.

In any case, this PR would need a overall remake because of all the changes since the last commit. So it is probably better to close it and open a new one for the uncertainties. (after fixing the `rebin()` method)

evandromr commented:
Deprecated by #141
